rust/kernel/chrdev.rs

```diff
@@ -149,12 +197,14 @@ unsafe impl<const N: usize> Sync for Registration<{ N }> {}
 impl<const N: usize> Drop for Registration<{ N }> {
     fn drop(&mut self) {
         if let Some(inner) = self.inner.as_mut() {
-            // SAFETY: Calling unsafe functions, `0..inner.used` of
-            // `inner.cdevs` are initialized in `Registration::register`.
+            // Replicate kernel C behaviour: drop [`Cdev`]s before calling
+            // [`bindings::unregister_chrdev_region`].
+            for i in 0..inner.used {
+                inner.cdevs[i].take();
+            }
+            // SAFETY: [`self.inner`] is Some, so [`inner.dev`] was previously
+            // created using [`bindings::alloc_chrdev_region`].
             unsafe {
-                for i in 0..inner.used {
-                    bindings::cdev_del(inner.cdevs[i].as_mut_ptr());
-                }
                 bindings::unregister_chrdev_region(inner.dev, N.try_into().unwrap());
```

Comment:
Should we add a comment here about why the `unwrap()` can never panic? E.g.

```rust
            // N.try_into() returned Some in [`register`], and N is constant, so it will return
            // Some here too, which ensures [`unwrap`] will not panic.
            bindings::unregister_chrdev_region(inner.dev, N.try_into().unwrap());
```

Reply:
Yes, please!

In fact, I think we could start `// PANIC:` and `// NOPANIC:` annotations nearby `unwrap()`, `expect()`, etc.
  - The former would annotate those that we believe may actually panic -- so it needs to be a justification of why the outer function was not made fallible to begin with.
  - The latter would annotate those that we believe cannot panic -- so it needs to justify why that is so. i.e. this is the category that could have potentially been `unwrap_unchecked()`.

At the moment, we have likely too many places due to allocations, so it can wait since otherwise we would be flooded with this comments; but later on we really should not have too many of these comments, which makes it worth the noise to have them, I think.

Thoughts?

Reply:
Agreed, the whole fallible allocations thing is still in flux, so adding these annotations all over the place probably won't be productive. So let's defer until later.

That said, I'm not sure, but I think we can eliminate the `unwrap` here:
```rust
let mut inner = this.inner.as_mut().unwrap();
```
simply by folding the `alloc_chrdev_region` into `new`. In that case we wouldn't need `self.inner` to be an `Option<RegistrationInner>`, it could simply be a `RegistrationInner`. Things become simpler, and there's one fewer potential unwrap panic.

Not sure if it's worth doing a follow-up PR for this (if it does indeed work)? Maybe there are more important fish to fry at this point?

Reply:
PS I'm not sure what the CI failure is about? The CI worked fine in my cloned repo. Same git id. https://github.com/TheSven73/linux/actions/runs/778067926

Reply:
Hmm... I just re-triggered them. `kobject_add` appears there, but looks it is not coming from us:

    [    0.628134] RIP: 0010:sysfs_create_file_ns+0x73/0x90
    ...
    [    0.628165]  kobject_add_internal+0x21a/0x3c0
    [    0.628167]  kobject_add+0x88/0xe0
    [    0.628170]  add_sysfs_fw_map_entry+0x61/0x70
    [    0.628173]  firmware_memmap_init+0x1f/0x2a
    [    0.628175]  ? firmware_map_add_early+0x57/0x57
    [    0.628178]  do_one_initcall+0xb3/0x240

Reply:
> Agreed, the whole fallible allocations thing is still in flux, so adding these annotations all over the place probably won't be productive. So let's defer until later.

We could still start annotating the ones unrelated to allocations, though (nevertheless, I am not asking to be added here, since we still need to discuss whether to do it or not).

Reply:
> Not sure if it's worth doing a follow-up PR for this (if it does indeed work)? Maybe there are more important fish to fry at this point?

Never mind, I'm running into borrow checker issue due to the `Pin`, a concept which I don't understand well enough yet. I'm going to leave this well alone :)

Reply:
`Pin` is indeed a bit tricky due to limitations of the language; the concept itself is easy. Hopefully the language will improve the ergonomics. Also the docs around it could be improved a bit.